File: ZIG/raytracer-zig/src/shapes/plane.zig
const std = @import("std");
const print = @import("std").debug.print;
const tMod = @import("../types.zig");
// const canvas = @import("../canvas.zig");
const utils = @import("../utils.zig");

const shapes = @import("shapes.zig");
const matrix = @import("../matrix.zig");
const mat_module = @import("../material.zig");
const intersection_mod = @import("intersections.zig");
const local_hits_mod = @import("local_hits.zig");

pub const Mat4 = matrix.Mat4;
const ShapeHeader = @import("shape_header.zig").ShapeHeader;

const S = tMod.S;
const Ray = tMod.Ray;
const Scalar = tMod.Scalar;
const Shape = shapes.Shape;
const Tuple = tMod.Tuple;
const Matrix = matrix.Mat4;
const Intersection = intersection_mod.Intersection;
const Intersections = intersection_mod.Intersections;
const LocalIntersections = intersection_mod.LocalIntersections;
const point = tMod.Point;
const vector = tMod.Vector;
const approxEq = tMod.approxEq;
const log = utils.log;
const Material = mat_module.Material;
const LocalHits = local_hits_mod.LocalHits;
const local_hits = local_hits_mod.LocalHits;

var NEXT_PLANE_ID: std.atomic.Value(usize) = .{ .raw = 1 };

pub const Plane = struct {
    h: ShapeHeader,

    pub const PlaneHits = local_hits(1);

    pub fn init() Plane {
        const obj_id = NEXT_PLANE_ID.fetchAdd(1, .seq_cst);
        // log(@src(), "\nNext Plane Id:{}\n", .{obj_id});
        return .{
            .h = ShapeHeader.init(obj_id),
        };
    }

    pub inline fn id(self: *const Plane) usize {
        // log(@src(), "\nNext Plane Id:{}\n", .{self.h.object_id});
        return self.h.object_id;
    }

    // The logic to intersect a ray with a plane is the only other bit that
    // needs implementing, and it has four cases to consider:

    // 1. The ray is
    // parallel to the plane, and will thus never intersect it.

    // 2. The ray is coplanar with the plane, which is to say that the ray’s
    // origin is on the plane, and the ray’s direction is parallel to the
    // plane. You’re viewing the plane edge-on. In this case, every point on
    // the ray intersects the plane, resulting in an infinite number of
    // intersections. That’s unwieldy! But since a plane is infinitely thin,
    // it’s invisible when viewed like this, so we’ll assume the ray misses in
    // this case.

    // 3. The ray origin is above the plane.

    // 4. The ray origin is below the plane.

    pub fn local_intersect(localray: Ray) PlaneHits {
        if (@abs(localray.direction.y) < tMod.EPSILON) {
            return .{ .count = 0, .t = .{S(0)} };
        }
        const t1 = -localray.origin.y / localray.direction.y;
        return .{ .count = 1, .t = .{t1} };
    }

    /// ---
    /// Recieve a local ray by that has been processed by applying the
    /// inverse of the sphere transform. Use the local rays origin and
    /// direction to compute the Intersections.
    /// ---
    pub fn intersect(self: *const Plane, ray: Ray) PlaneHits {
        const localray = Ray{
            .origin = self.h.transformed_m_inv.mulT(ray.origin),
            .direction = self.h.transformed_m_inv.mulT(ray.direction),
        };

        return local_intersect(localray);
    }

    /// ---
    /// NOTE: returns the plane-specific normal
    /// ---
    pub fn local_normal_at(self: *const Plane, local_point: Tuple) Tuple {
        _ = self;
        _ = local_point;
        return vector(0, 1, 0);
    }

    /// ---
    /// Setting the transform matrix.
    /// NOTE: Also computes the inverse as a side effect...
    /// ---
    pub fn set_transform(self: *Plane, m: matrix.Mat4) void {
        return self.h.set_transform(m);
    }

    pub fn transform(self: *const Plane) matrix.Mat4 {
        return self.h.transformed_m;
    }

    pub fn inverse(self: *const Plane) matrix.Mat4 {
        return self.h.transformed_m_inv;
    }

    pub fn reset_id() void {
        const obj_id = NEXT_PLANE_ID.load(.seq_cst);
        log(@src(), "\nNext Plane Id:{}\n", .{obj_id});
        NEXT_PLANE_ID.store(1, .seq_cst);
    }
};

test "Make sure it works" {
    try std.testing.expect(3 == 3);
}

test "Chap9 -The normal of a plane is constant everywhere" {
    const pl = Plane.init();
    const n1 = pl.local_normal_at(point(0, 0, 0));
    const n2 = pl.local_normal_at(point(10, 0, -10));
    const n3 = pl.local_normal_at(point(-5, 0, 150));
    try std.testing.expect(pl.h.transformed_m.equals(&Mat4.identity()));
    try std.testing.expect(n1.equals(vector(0, 1, 0)));
    try std.testing.expect(n2.equals(vector(0, 1, 0)));
    try std.testing.expect(n3.equals(vector(0, 1, 0)));
}

test "Chap9 -Intersect with a ray parallel to the plane" {
    const pl = Plane.init();
    _ = pl;
    const r = Ray.init(point(0, 10, 0), vector(0, 0, 1));
    const xs = Plane.local_intersect(r);
    try std.testing.expect(xs.count == 0);
}

test "Chap9 -Intersect with a coplanar ray" {
    const r = Ray.init(point(0, 0, 0), vector(0, 0, 1));
    const xs = Plane.local_intersect(r);
    try std.testing.expect(xs.count == 0);
}

test "Chap9 -A ray intersecting a plane from above" {
    const r = Ray.init(point(0, -1, 0), vector(0, 1, 0));
    const xs = Plane.local_intersect(r);
    try std.testing.expect(xs.count == 1);
    try std.testing.expect(approxEq(S(1), xs.items()[0]));
    // Can not really do the id check based on local_intersect
    // since the object id is stored in the header and provided
    // by the Intersect.
}

File: RUST/rtc-rust/src/shapes/plane.rs
//! Plane definition
//!
//! An infinite flat surface extending in the xz plane by default.
//! The normal of a plane is constant everywhere.
//!

use crate::intersection::{Intersection, Intersections};
use crate::ray::Ray;
use crate::shape::{Shape, ShapeData};
use crate::tuple::Tuple;

// #[derive(Debug, Clone, Copy)]
#[derive(Debug, Clone)]
pub struct Plane {
    pub data: ShapeData,
}

impl Plane {
    pub fn new() -> Self {
        Self {
            data: ShapeData::new(),
        }
    }
}

impl Default for Plane {
    fn default() -> Self {
        Self::new()
    }
}

impl Shape for Plane {
    fn data(&self) -> &ShapeData {
        &self.data
    }

    fn data_mut(&mut self) -> &mut ShapeData {
        &mut self.data
    }

    ///
    /// The normal of a plane is constant everywhere
    ///
    fn local_normal_at(&self, _point: Tuple) -> Tuple {
        Tuple::vector(0.0, 1.0, 0.0)
    }

    fn local_intersect(&self, ray: &Ray) -> Intersections {

        if ray.direction.y.abs() < crate::math::EPSILON {
            return Intersections::new();
        }

        let mut i = Intersections::new();
        let t = -ray.origin.y / ray.direction.y;
        let xs = Intersection::new(t, self.id());
        i.push(xs);
        i
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::math::approx_eq;
    use crate::ray::Ray;
    use crate::tuple::Tuple;

    /// Chap 9 - The normal of a plane is constant everywhere
    #[test]
    fn test_chap_9_1() -> Result<(), String> {
        let p = Plane::new();
        let n1 = p.local_normal_at(Tuple::point(0.0, 0.0, 0.0));
        let n2 = p.local_normal_at(Tuple::point(10.0, 0.0, -10.0));
        let n3 = p.local_normal_at(Tuple::point(-5.0, 0.0, 150.0));
        let expected = Tuple::vector(0.0, 1.0, 0.0);
        let chk = n1.approx_eq(expected) && n2.approx_eq(expected) && n3.approx_eq(expected);
        if chk {
            Ok(())
        } else {
            Err("The normal of a plane is constant everywhere".into())
        }
    }

    /// Chap 9 - Intersect with a ray parallel to the plane
    #[test]
    fn test_chap_9_2() -> Result<(), String> {
        let p = Plane::new();
        let r = Ray::new(Tuple::point(0.0, 10.0, 0.0), Tuple::vector(0.0, 0.0, 1.0));
        let xs = p.local_intersect(&r);
        let chk = xs.is_empty();
        if chk {
            Ok(())
        } else {
            Err("Intersect with a ray parallel to the plane".into())
        }
    }

    /// Chap 9 - Intersect with a coplanar ray
    #[test]
    fn test_chap_9_3() -> Result<(), String> {
        let p = Plane::new();
        let r = Ray::new(Tuple::point(0.0, 0.0, 0.0), Tuple::vector(0.0, 0.0, 1.0));
        let xs = p.local_intersect(&r);
        let chk = xs.is_empty();
        if chk {
            Ok(())
        } else {
            Err("Intersect with a coplanar ray".into())
        }
    }

    /// Chap 9 - A ray intersecting a plane from above
    #[test]
    fn test_chap_9_4() -> Result<(), String> {
        let p = Plane::new();
        let r = Ray::new(Tuple::point(0.0, 1.0, 0.0), Tuple::vector(0.0, -1.0, 0.0));
        let xs = p.local_intersect(&r);
        let chk = xs.count() == 1;
        let chk = chk && approx_eq(xs[0].t, 1.0);
        let chk = chk && xs[0].object_id == p.id();
        if chk {
            Ok(())
        } else {
            Err("A ray intersecting a plane from above".into())
        }
    }

    /// Chap 9 - A ray intersecting a plane from below
    #[test]
    fn test_chap_9_5() -> Result<(), String> {
        let p = Plane::new();
        let r = Ray::new(Tuple::point(0.0, -1.0, 0.0), Tuple::vector(0.0, 1.0, 0.0));
        let xs = p.local_intersect(&r);
        let chk = xs.count() == 1;
        let chk = chk && approx_eq(xs[0].t, 1.0);
        let chk = chk && xs[0].object_id == p.id();
        if chk {
            Ok(())
        } else {
            Err("A ray intersecting a plane from below".into())
        }
    }
}
